WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,273 @@
+        /// The initial sweep is performed after a delay to avoid competing with
+        /// other subsystems for the resources.
+        var initialSweepDelay: TimeInterval = 10
+
+        /// Initializes the configuration.
+        ///
+        /// - parameter sizeLimit: Size limit in bytes. `100 Mb` by default.
+        public init(sizeLimit: Int = 1024 * 1024 * 100) {
+            self.sizeLimit = sizeLimit
+        }
+    }
+
+    /// Creates a cache instance with a given `name`. The cache creates a directory
+    /// with the given `name` in a `.cachesDirectory` in `.userDomainMask`.
+    public init(name: String, configuration: Configuration = .init()) {
+        let cachesURL = DiskCache.getCachesURL()
+        self.init(url: cachesURL.appendingPathComponent(name, isDirectory: true), configuration: configuration)
+    }
+
+    private static func getCachesURL() -> URL {
+        guard let cachesURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            DDLogError("Failed to instantiate cache. Caches directory not available")
+            return URL(fileURLWithPath: "/dev/null") // This should never happen
+        }
+        return cachesURL
+    }
+
+    /// Creates a cache instance with a given root URL.
+    public init(url: URL, configuration: Configuration = .init()) {
+        self.rootURL = url
+        self.configuration = configuration
+
+        do {
+            try FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+        } catch {
+            DDLogError("Failed to creates cache root directory at: \(url) with error: \(error)")
+        }
+        DispatchQueue.main.asyncAfter(deadline: .now() + configuration.initialSweepDelay) { [weak self] in
+            Task {
+                await self?.performAndScheduleNextSweep()
+            }
+        }
+    }
+
+    // MARK: Codable
+
+    public func getValue<T: Decodable>(
+        _ type: T.Type,
+        forKey key: String,
+        decoder: JSONDecoder = JSONDecoder()
+    ) async -> T? {
+        guard let data = getData(forKey: key) else { return nil }
+        return try? decoder.decode(type, from: data)
+    }
+
+    public func setValue<T: Encodable>(
+        _ value: T,
+        forKey key: String,
+        encoder: JSONEncoder = JSONEncoder()
+    ) async {
+        guard let data = try? encoder.encode(value) else { return }
+        setData(data, forKey: key)
+    }
+
+    public func removeValue(forKey key: String) {
+        removeData(forKey: key)
+    }
+
+    // MARK: Codable (Closures)
+
+    public nonisolated func getValue<T: Decodable>(
+        _ type: T.Type,
+        forKey key: String,
+        decoder: JSONDecoder = JSONDecoder(),
+        _ completion: @escaping (T?) -> Void
+    ) {
+        Task {
+            let value = await getValue(type, forKey: key, decoder: decoder)
+            completion(value)
```

Comment:
As a rule, I'd suggest only using `Task` blocks from view controllers and similar – structured concurrency has a surprisingly low limit on how many operations can be in-flight at once, and IMHO using this pattern in a library increases the risk of deadlock.

The typical pattern would be to reverse this – to run the code on a private dispatch queue, then wrap it in `withUnsafeThrowingContinuation` for use with async/await if the caller prefers that.

Reply:
I wasn't sure about adding it. Yes, it's probably best to leave it to the caller as you suggested. But It's straight from the WWDC talk on Macros, except it uses detach, which I'm not sure is needed.

<img width="1616" alt="Screenshot 2023-07-11 at 7 25 04 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/50b3df0e-d0a7-4cbd-8ba0-b26c0107257c">

>  low limit on how many operations can be in-flight at once, and IMHO using this pattern in a library increases the risk of deadlock.

Can you please expand on that? My understanding of Swift Concurrency is still fairly limited, and it's in a pre-release state in many ways. But based on my testing, tasks are extremely cheap and can not lead to a deadlock or a thread explosion. There can be a virtually limitless number of suspended in-flight tasks. That's probably the main selling point.

Reply:
`DiskCache` is now a simple `class`, not an `actor`. This problem no longer exists.

Reply:
I see that the `async` stuff was removed, but consider the case of some `UIViewController` using `Task` to wrap a call into an `ObjectRepository` method with a callback, and that method uses this cache under the hood. I don't know for sure that this is still the case, but at one point the simulator only allowed a single in-flight `Task` at once – IINM this situation would result in a deadlock, and there's no way to avoid it because we can't guarantee that nobody will ever use the callback-based implementation from within a `Task` block.

Reply:
Yeah, it was too contentious, so I decided to remove it. I generally don't trust the new concurrency system in Swift yet. I think it's still in a pre-release state. Folks have been trying to use it in production, and they immediately run into issues. For example, [this](https://forums.swift.org/t/async-await-is-it-possible-to-start-a-task-on-mainactor-synchronously/52862) has been a huge limiting factor to designing anything that plays well with UIKit.

> the simulator only allowed a single in-flight Task

There were a few issues like this during the early beta, especially with a back-ported implementation. I think it was sorted. But ditto it's in a pre-release state even now.